### PR TITLE
Query: Adds LINQ support for Multi-key Group By translation

### PR DESCRIPTION
**Pull request by leminh98**
# Pull Request Template

## Description

This PR features the new syntax for multi-key group by queries, allowing for grouping of multiple properties. 

Full syntax: 
```CSharp

// Function Signature
public static System.Collections.Generic.IEnumerable<System.Linq.IGrouping<TKey,TElement>> GroupBy<TSource,TKey,TElement> (this System.Collections.Generic.IEnumerable<TSource> source, Func<TSource,TKey> keySelector, Func<TSource,TElement> elementSelector);

// Usage
Enumerables.GroupBy( groupByKeys => keySelectiorLambda, (keys, values) => valueSelectorLambda);
```

Example
- Grouping by single keys and projecting single value
 ```Csharp
Enumerable.GroupBy( /*keySelector*/ k => k.Id,  /*valueSelector*/ (key, values) => values.Count());
```

- Grouping by single keys and projecting multiple values
 ```Csharp
Enumerable.GroupBy( 
               /*keySelector*/ k => k.Id, 
               /*valueSelector*/ (key, values) => new
                {
                    idField = key.key1,
                    count = values.Count(),
                });
```

- Grouping by multiple keys and projecting single value
 ```Csharp
Enumerable.GroupBy( 
               /*keySelector*/  k => new
                  {
                      key1 = k.Id,
                      key2 = k.Int
                  } , 
               /*valueSelector*/ (key, values) => values.Count());
```

- Grouping by multiple keys and projecting multiple values
```Csharp
Enumerable.GroupBy(
/*keySelector*/
              k => new
                  {
                      key1 = k.Id,
                      key2 = k.Int
                  } , 
/*valueSelector*/
                (key, values) => new
                {
                    idField = key.key1,
                    count = values.Count(),
                })
```
## Type of change

Please delete options that are not relevant.

- [] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] This change requires a documentation update

## Closing issues
closes ##1202

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs [hunk 1]
```diff
@@ -1793,44 +1784,121 @@ private static Collection VisitGroupBy(Type returnElementType, ReadOnlyCollectio
...
-
-            // The alias for the key in the value selector lambda is the first arguemt lambda - we bound it to the parameter expression, which already has substitution
-            ParameterExpression valueSelectorKeyExpressionAlias = Utilities.GetLambda(arguments[2]).Parameters[0];
-            context.GroupByKeySubstitution.AddSubstitution(valueSelectorKeyExpressionAlias, parameterExpression/*Utilities.GetLambda(arguments[1]).Body*/);
-
-            // Translate the body of the value selector lambda
+            LambdaExpression keySelectorLambda = Utilities.GetLambda(arguments[1]);
+
+            //// Current GroupBy doesn't allow subquery, so we need to visit non subquery scalar lambda
+            //SqlScalarExpression keySelectorFunc = ExpressionToSql.VisitNonSubqueryScalarLambda(keySelectorLambda, context);
```
Comment (adityasa) [inline Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs:1793]: Remove if not required.

File: Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs [hunk 2]
```diff
@@ -1793,44 +1784,121 @@ private static Collection VisitGroupBy(Type returnElementType, ReadOnlyCollectio
...
+
+                        context.CurrentQuery = context.CurrentQuery.AddGroupByClause(groupby, context);
+
+                        ParameterExpression parameterExpression = context.GenerateFreshParameter(returnElementType, keySelectorFunc.ToString(), includeSuffix: false);
+                        Binding binding = new Binding(parameterExpression, collection.inner, isInCollection: false, isInputParameter: true);
+
+                        context.CurrentQuery.GroupByParameter.Add(binding);
+                        // The alias for the key in the value selector lambda is the first arguemt lambda - we bound it to the parameter expression, which already has substitution
+                        ParameterExpression valueSelectorKeyExpressionAlias = Utilities.GetLambda(arguments[2]).Parameters[0];
+                        context.GroupByKeySubstitution.AddSubstitution(valueSelectorKeyExpressionAlias, parameterExpression/*Utilities.GetLambda(arguments[1]).Body*/);
```
Comment (adityasa) [inline Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs:1822]: >/*Utilities.GetLambda(arguments[1]).Body*/ [](http://example.com/codeflow?start=123&length=42)

Remove if not required.

File: Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs [hunk 3]
```diff
@@ -1793,44 +1784,121 @@ private static Collection VisitGroupBy(Type returnElementType, ReadOnlyCollectio
...
+
+                        break;
+                    }
+                case ExpressionType.New:
+                    {
+                        NewExpression newExpression = (NewExpression)keySelectorLambda.Body;
+
+                        if (newExpression.Members == null)
+                        {
+                            throw new DocumentQueryException(ClientResources.ConstructorInvocationNotSupported);
```
Comment (adityasa) [inline Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs:1832]: Do you have coverage for this case? If so, can you please share a pointer?

File: Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs [hunk 4]
```diff
@@ -1793,44 +1784,121 @@ private static Collection VisitGroupBy(Type returnElementType, ReadOnlyCollectio
...
+
+                        context.CurrentQuery.GroupByParameter.Add(binding);
+                        // The alias for the key in the value selector lambda is the first arguemt lambda - we bound it to the parameter expression, which already has substitution
+                        ParameterExpression valueSelectorKeyExpressionAlias = Utilities.GetLambda(arguments[2]).Parameters[0];
+                        context.GroupByKeySubstitution.AddSubstitution(valueSelectorKeyExpressionAlias, parameterExpression/*Utilities.GetLambda(arguments[1]).Body*/);
+
+                        break;
+                    }
+                case ExpressionType.New:
+                    {
```
Comment (adityasa) [inline Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs:1827]: There is quite a bit of code common between this case and previous cases. Is it possible to move the common code somewhere, for e.g. at the end of the switch...case?

File: Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs [hunk 5]
```diff
@@ -1793,44 +1784,121 @@ private static Collection VisitGroupBy(Type returnElementType, ReadOnlyCollectio
...
+
+                        // Bind the alias
+                        ParameterExpression parameterExpression = context.GenerateFreshParameter(returnElementType, keySelectorFunctions.ToString(), includeSuffix: false);
+                        Binding binding = new Binding(parameterExpression, collection.inner, isInCollection: false, isInputParameter: true);
+
+                        context.CurrentQuery.GroupByParameter.Add(binding);
+                        // The alias for the key in the value selector lambda is the first arguemt lambda - we bound it to the parameter expression, which already has substitution
+                        ParameterExpression valueSelectorKeyExpressionAlias = Utilities.GetLambda(arguments[2]).Parameters[0];
+                        context.GroupByKeySubstitution.AddSubstitution(valueSelectorKeyExpressionAlias, parameterExpression/*Utilities.GetLambda(arguments[1]).Body*/);
+
```
Comment (adityasa) [inline Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs:1865]: Seems common with previous switch...case

File: Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs [hunk 6]
```diff
@@ -1793,44 +1784,121 @@ private static Collection VisitGroupBy(Type returnElementType, ReadOnlyCollectio
...
+                        if (newExpression.Members == null)
+                        {
+                            throw new DocumentQueryException(ClientResources.ConstructorInvocationNotSupported);
+                        }
+
+                        //bind the parameters in the value selector to the current input
+                        foreach (ParameterExpression par in Utilities.GetLambda(arguments[1]).Parameters)
+                        {
+                            context.PushParameter(par, context.CurrentSubqueryBinding.ShouldBeOnNewQuery);
+                        }
```
Comment (adityasa) [inline Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs:1839]: Same as previous cases. Can this be moved out of switch...case?

**Comment by adityasa** [comment]
>             bool ignoreOrderingForAnonymousTypeObject = false,

Please rename to just ignoreOrder

---
Refers to: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Linq/LinqTestsCommon.cs:686 in 7c8ff08. [](commit_id = 7c8ff0816ee00a996849561074cf057e77bece89, deletion_comment = False)

File: Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs (hunk 3 above)
Comment (leminh98) [reply Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs:1832]: added new tests to cover this case

File: Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs (hunk 6 above)
Comment (leminh98) [reply Microsoft.Azure.Cosmos/src/Linq/ExpressionToSQL.cs:1839]: this case has different push so I kept the for loop inside. Tried and moved the binding later out of the switch

**Review by adityasa — DISMISSED** [review]
:shipit:

**Review by adityasa — APPROVED** [review]
:shipit:

**Review by Maya-Painter — APPROVED** [review]

**Comment by benjaminmillhouse** [comment]
Hello,
This feature is working great - however, I've noticed that we cannot use a concrete class in the Projection portion of the call... (valueSelector). If we try to use one instead of an anonymous object, we receive the error: `Microsoft.Azure.Cosmos.Linq.DocumentQueryException: Expression with NodeType 'MemberInit' is not supported.`

I'm just curious if that is expected behavior?  I have been able to use concrete classes over anonymous objects in .Select() operations before, this seems specific to the new GroupBy functionality.

Thanks!

**Comment by sacgaikwad** [comment]
hello,

I was looking for multikey grouping, however when I pass the multiple keys it give error:
**Microsoft.Azure.Cosmos.Linq.DocumentQueryException: 'Only path expressions are supported for SelectMany.'**

**Method signature:**
public async Task<IEnumerable<TResult>> QueryAsync<T, TKey, TResult>(Expression<Func<T, bool>> filter, Expression<Func<T, TKey>> keySelector, Expression<Func<TKey, IEnumerable<T>, TResult>> resultSelector, CancellationToken cancellationToken = default)

**Exception at :**
     var query = container.GetItemLinqQueryable<T>().Where(filter).GroupBy(keySelector, resultSelector).ToFeedIterator();

**input:**
 Expression<Func<ApiMetrics, dynamic>> keySelector = query =>
new
{
    key1 = query.CompanyId,
    key2 = query.ProductType
};


Can someone help here?

**Comment by Amit-BC** [comment]
> hello folks,
> 
> I was looking for multikey grouping, however when I pass the multiple keys it give error: **Microsoft.Azure.Cosmos.Linq.DocumentQueryException: 'Only path expressions are supported for SelectMany.'**
> 
> **Method signature:** public async Task<IEnumerable> QueryAsync<T, TKey, TResult>(Expression<Func<T, bool>> filter, Expression<Func<T, TKey>> keySelector, Expression<Func<TKey, IEnumerable, TResult>> resultSelector, CancellationToken cancellationToken = default)
> 
> **Exception at :** var query = container.GetItemLinqQueryable().Where(filter).GroupBy(keySelector, resultSelector).ToFeedIterator();
> 
> **input:** Expression<Func<ApiMetrics, dynamic>> keySelector = query => new { key1 = query.CompanyId, key2 = query.ProductType };
> 
> Can someone help here?

is there a solution for this? I get the same error

**Comment by Maya-Painter** [comment]
@benjaminmillhouse @sacgaikwad @Amit-BC If you are still receiving these errors in version 3.47.0 please open a new issue so we can investigate further.

**Comment by benjaminmillhouse** [comment]
> @benjaminmillhouse @sacgaikwad @Amit-BC If you are still receiving these errors in version 3.47.0 please open a new issue so we can investigate further.

The issue I reported seems to be resolved now.  Thanks!

**Comment by next-phil** [comment]
Does this document need updating to mention Group By being available now?

https://learn.microsoft.com/en-us/cosmos-db/query/linq-to-sql#supported-linq-operators

**Pull request merged**